### ISSUE TITLE
linter: no-useless-spread fails with iterators + map

### ISSUE THREAD

**Issue by posva** [oxc-project/oxc#11638]
### What version of Oxlint are you using?

1.1.0

### What command did you run?

`oxlint`

### What does your `.oxlintrc.json` config file look like?

```jsonc
// noconfig
```

### What happened?

It fails with this code:

```ts
function* find() {
  yield { data: 1 }
  yield { data: 2 }
}

expect([...find().map((e) => e.data)]).toEqual([1, 2]) 
```

**Comment by camc314** [comment]
Just a note, this is basically impossible to fix without type inference information as we have no idea whether `find()` returns an array or yields a value.

**Comment by posva** [comment]
Ah I see. Is the rule just flawed in modern JavaScript then?

**Comment by camc314** [comment]
> Ah I see. Is the rule just flawed in modern JavaScript then?

Not really? there's plenty of cases described in the rule's docs that would cause an unneccessary array clone/allocation, the rule would still correctly report these cases. My point about this false positive being tricky to fix referred specifically to the lack of type information alowing us to identify whether this spread is valid or not.

**Comment by posva** [comment]
If it cannot be applied anymore all the time, it’s flawed (by definition), that’s what I mean. Of course there are cases where you don’t want the extra copy

I haven’t checked the source of the rule but maybe, as you said, it might not be possible at all to have a rule that works all the time. In that case, I think it would be better to have it disabled by default to avoid false negatives

**Comment by posva** [comment]
@Boshen This is still an issue: the rule shouldn't be on by default because it breaks valid code: it transforms this:

```ts
function* find() {
  yield { data: 1 }
  yield { data: 2 }
}

expect([...find().map((e) => e.data)]).toEqual([1, 2]) 
```

into

```ts
function* find() {
  yield { data: 1 }
  yield { data: 2 }
}

expect(find().map((e) => e.data)).toEqual([1, 2]) 
```

I think the rule being opt-in would be a benefit
